python-sdk/src/astro/databricks/delta.py

```diff
@@ -312,3 +302,106 @@ def load_pandas_dataframe_to_table(
+        :param target_table: Contains the destination table in which the rows will be merged
+        :param source_to_target_columns_map: Dict of target_table columns names to source_table columns names
+        :param target_conflict_columns: List of cols where we expect to have a conflict while combining
+        :param if_conflicts: The strategy to be applied if there are conflicts.
+        """
+        statement, params = self._build_merge_sql(
+            source_table=source_table,
+            target_table=target_table,
+            source_to_target_columns_map=source_to_target_columns_map,
+            target_conflict_columns=target_conflict_columns,
+            if_conflicts=if_conflicts,
+        )
+        self.run_sql(sql=statement, parameters=params)
+
+    def _build_merge_sql(
+        self,
+        source_table: BaseTable,
+        target_table: BaseTable,
+        source_to_target_columns_map: dict[str, str],
+        target_conflict_columns: list[str],
+        if_conflicts: MergeConflictStrategy = "exception",
+    ):
+        """Build the SQL statement for Merge operation"""
+        # TODO: Simplify this function
+        source_table_name = source_table.name
+        target_table_name = target_table.name
+
+        source_cols = source_to_target_columns_map.keys()
+        target_cols = source_to_target_columns_map.values()
+
+        (
+            source_table_identifier,
+            source_table_param,
+        ) = self.get_sqlalchemy_template_table_identifier_and_parameter(source_table, "source_table")
+
+        (
+            target_table_identifier,
+            target_table_param,
+        ) = self.get_sqlalchemy_template_table_identifier_and_parameter(target_table, "target_table")
+
+        statement = (
+            f"merge into {target_table_identifier} as `target_table` using {source_table_identifier} as `source_table` "
+            + "on {merge_clauses}"
+        )
+
+        merge_target_dict = {
+            f"`target_table`.`{x}`": f"`target_table`.`{x}`" for i, x in enumerate(target_conflict_columns)
+        }
+        merge_source_dict = {
+            f"`source_table`.`{x}`": f"`source_table`.`{x}`" for i, x in enumerate(target_conflict_columns)
+        }
+        statement = statement.replace(
+            "{merge_clauses}",
+            " AND ".join(f"{k}={v}" for k, v in zip(merge_target_dict.keys(), merge_source_dict.keys())),
+        )
+
+        values_to_check = [target_table_name, source_table_name]
+        values_to_check.extend(source_cols)
+        values_to_check.extend(target_cols)
+        if if_conflicts == "update":
+            statement += " when matched then UPDATE SET {merge_vals}"
+            merge_statement = ",".join(
+                [f"target_table.{t} = source_table.{s}" for s, t in source_to_target_columns_map.items()]
+            )
+            statement = statement.replace("{merge_vals}", merge_statement)
```

Comment:
```suggestion
            merge_statement = ",".join(
                [f"target_table.{t} = source_table.{s}" for s, t in source_to_target_columns_map.items()]
            )
            statement += f" when matched then UPDATE SET {merge_statement}"
```
This can be simplified using f-string.

We should not use `replace` for formatting strings. This is really error-prone! Using `format` should do the same job better.

Reply:
This is a great call. I'm going to make a separate PR to do do the same with the SNOWFLAKE queries.